service/src/java/org/apache/hive/service/cli/thrift/ThriftBinaryCLIService.java

```diff
@@ -105,4 +140,79 @@ public void run() {
     }
   }
 
+  // Custom authentication class with Kerberos thread
+  private static void startCustomWithKerberos(
+   final HiveConf hiveConf,
+   final ThriftCLIService service,
+   final String hiveHost,
+   TServer customKrbServer) throws Exception {
+
+   try {
+     int minThreads = hiveConf.getIntVar(ConfVars.HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_MIN_WORKER_THREADS);
+     int maxThreads = hiveConf.getIntVar(ConfVars.HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_MAX_WORKER_THREADS);
+
+     // custom authenticatino class with Kerberos Server thread pool
+     String threadPoolName = "HiveServer2-custom-with-Krb-Handler-Pool";
+     ExecutorService executorService = new ThreadPoolExecutor(minThreads, maxThreads,
+         service.workerKeepAliveTime, TimeUnit.SECONDS, new SynchronousQueue<Runnable>(),
+         new ThreadFactoryWithGarbageCleanup(threadPoolName));
+
+     int customPortNum;
+     String portString = System.getenv("HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_PORT");
+     if (portString != null) {
+       customPortNum = Integer.valueOf(portString);
+     } else {
+       customPortNum = hiveConf.getIntVar(ConfVars.HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_PORT);
+     }
+
+     HiveAuthFactory hiveAuthFactory = new HiveAuthFactory(hiveConf);
+     TTransportFactory transportFactory = hiveAuthFactory.getAuthPlainTransFactory();
+     TProcessorFactory processorFactory = hiveAuthFactory.getAuthProcFactory(service);
+     TServerSocket customKrbSocket = null;
+
+     if (!hiveConf.getBoolVar(ConfVars.HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_SSL_USED)) {
+       customKrbSocket = HiveAuthFactory.getServerSocket(hiveHost, customPortNum);
+     } else {
+       List<String> sslVersionBlacklist = new ArrayList<String>();
+       for (String sslVersion : hiveConf.getVar(ConfVars.HIVE_SSL_PROTOCOL_BLACKLIST).split(",")) {
+         sslVersionBlacklist.add(sslVersion);
+       }
+       String keyStorePath = hiveConf.getVar(ConfVars.HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_SSL_KEYSTORE_PATH).trim();
+       if (keyStorePath.isEmpty()) {
+         throw new IllegalArgumentException(ConfVars.HIVE_SERVER2_KERBEROS_CUSTOM_AUTH_SSL_KEYSTORE_PATH.varname +
+         " Not configured for SSL connection");
```

Comment:
you are configured for SSL_USED to get this far.  This error should say "empty SSL keystore path" or something similar to that